pkg/compute/serve.go

```diff
@@ -0,0 +1,287 @@
+func installViceroy(progress text.Progress, versioner update.Versioner, latest semver.Version, bin string) error {
+	progress.Step("Fetching latest viceroy release...")
+
+	tmp, err := versioner.Download(context.Background(), latest)
+	if err != nil {
+		progress.Fail()
+		return fmt.Errorf("error downloading latest viceroy release: %w", err)
+	}
+
+	if err := os.Rename(tmp, bin); err != nil {
+		if err := filesystem.CopyFile(tmp, bin); err != nil {
+			progress.Fail()
+			return fmt.Errorf("error moving latest viceroy binary in place: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// updateViceroy checks if the currently installed version is out-of-date and
+// downloads the latest release from GitHub.
+func updateViceroy(progress text.Progress, version string, out io.Writer, versioner update.Versioner, latest semver.Version, bin string) error {
+	progress.Step("Checking installed viceroy version...")
+
+	var installedViceroyVersion string
+
+	viceroyError := errors.RemediationError{
+		Inner:       fmt.Errorf("viceroy version not found"),
+		Remediation: errors.BugRemediation,
+	}
+
+	// version output has the expected format: `viceroy 0.1.0`
+	segs := strings.Split(version, " ")
+
+	if len(segs) < 2 {
+		return viceroyError
+	}
+
+	installedViceroyVersion = segs[1]
+
+	if installedViceroyVersion == "" {
+		return viceroyError
+	}
+
+	current, err := semver.Parse(installedViceroyVersion)
+	if err != nil {
+		progress.Fail()
+
+		return errors.RemediationError{
+			Inner:       fmt.Errorf("error reading current version: %w", err),
+			Remediation: errors.BugRemediation,
+		}
+	}
+
+	if latest.GT(current) {
+		text.Output(out, "Current viceroy version: %s", current)
+		text.Output(out, "Latest viceroy version: %s", latest)
+
+		progress.Step("Fetching latest viceroy release...")
+		tmp, err := versioner.Download(context.Background(), latest)
+		if err != nil {
+			progress.Fail()
+			return fmt.Errorf("error downloading latest viceroy release: %w", err)
+		}
+		defer os.RemoveAll(tmp)
+
+		progress.Step("Replacing viceroy binary...")
+
+		if err := os.Rename(tmp, bin); err != nil {
+			if err := filesystem.CopyFile(tmp, bin); err != nil {
+				progress.Fail()
+				return fmt.Errorf("error moving latest viceroy binary in place: %w", err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// local spawns a subprocess that runs the compiled binary.
+func local(bin string, out io.Writer, env string, verbose bool) error {
+	sig := make(chan os.Signal, 1)
+
+	signals := []os.Signal{
+		syscall.SIGINT,
+		syscall.SIGTERM,
+	}
+
+	signal.Notify(sig, signals...)
+
+	if env != "" {
+		env = "." + env
+	}
+	manifest := fmt.Sprintf("fastly%s.toml", env)
+	args := []string{"bin/main.wasm", "-C", manifest}
+
+	cmd := common.NewStreamingExec(bin, args, os.Environ(), verbose, out)
+
+	go func(sig chan os.Signal, cmd *common.StreamingExec) {
+		<-sig
+		signal.Stop(sig)
+
+		err := cmd.Signal(os.Kill)
+		if err != nil {
+			log.Fatal(err)
+		}
+	}(sig, cmd)
```

Comment:
This seems like good default behaviour and an oversight in the original implementation of `StreamingExec`, maybe we should consider moving this logic inside `streamingexec.Exec()` so all instances of it in the CLI benefit?